content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+
+|||codediff|rust
++ #[export_name = "add"]
+  pub fn add(left: usize, right: usize) -> usize {
+      left + right
+  }
+|||
+
+If you don’t intend to change the functions external name, you can also use `#[no_mangle]` instead, which will have the same effect. I feel like `no_mangle` hides the true intention of the developer here, so I tend to prefer the `export_name`. 
+
+Having given our `add` function an export name, we can compile the project again and inspect the resulting WebAssembly file:
+
+|||codediff|wasm
+  (module
++   (type (;0;) (func (param i32 i32) (result i32)))
++   (func $add (type 0) (param i32 i32) (result i32)
++     local.get 1
++     local.get 0
++     i32.add)
+    (table (;0;) 1 1 funcref)
+    (memory (;0;) 16)
+    (global $__stack_pointer (mut i32) (i32.const 1048576))
+    (global (;1;) i32 (i32.const 1048576))
+    (global (;2;) i32 (i32.const 1048576))
+    (export "memory" (memory 0))
++   (export "add" (func $add))
+    (export "__data_end" (global 1))
+    (export "__heap_base" (global 2)))
+|||
+
+This module is easy to run in Node, or Deno or even the browser:
+
+```js
+// Node
+const data = require("fs").readFileSync("./my_project.wasm");
+// Deno
+const data = await Deno.readFile("./my_project.wasm");
+// Web
+const data = await fetch("./my_project.wasm").then(r => r.arrayBuffer());
+
+const importObj = {
+};
+const {instance} = await WebAssembly.instantiate(data, importObj);
+instance.exports.add(40, 2) // returns 42
+```
+
+And suddenly, we have pretty much all the power of Rust at our fingertips to write WebAssembly. 
+
+> **Streaming:** On the web it is advisable to _not_ call `r.arrayBuffer()` and instead pass the `Response` straight to `WebAssembly.instantiateStreaming`. This allows the WebAssembly compiler to start compiling while the file is being downloaded.
+
+Special care needs to be taken with functions at the module boundary (i.e. the ones you call from JavaScript). It’s best to stick to types that map cleanly to [WebAssembly types] (like `i32` or `f64`). If you use higher-level types like arrays, slices, or even owned types like `String`, it will compile, but yield a unexpected function signature and will generally become harder to use. More on that later!
+
+### Importing
+
+One important part of WebAssembly is its sandbox. It ensures that the code running in the WebAssembly VM gets no access to anything in the host environment apart from the functions that were explicitly passed into the sandbox via the imports object.
+
+Let’s say we want to generate some random numbers in our Rust code. We could pull in the `rand` Rust crate, but why ship code for something if we can grab the existing implementation for free. To make that work, we need to declare that our WebAssembly module expects an import.
+
+|||codediff|rust
++ #[link(wasm_import_module = "Math")]
++ extern "C" {
++     #[link_name = "random"]
```

Comment:
You mention this below, but renaming via `link_name` is a relatively rarely used feature outside of bindgen-generated code, and IMO it's worth just leaving it out for simplicity. Or, perhaps, you can mention it in some side-note.